### PR TITLE
Feature/indexer/tvmaze

### PR DESCRIPTION
**Pull request by p0psicles**
## This pr has been initially been based on the TMDB pr. The TMDB pr will most likely be closed, while we will continue on this one.

Added tvmaze indexer.

Main reason for now, is to see what's still required to add additional indexers. Like for example, how usable are indexers that can't make use of the scene name exceptions. Or other services which require a thetvdb id.

**This PR can only be merged when TMDB has been merged to develop. And this PR has been rebased on dev!**

### Log:

This is a list to document what needs to be taken care of when implementing a new indexer. I'll update this as I move along.
- [x] Show search by name
- [x] Show search by id, indexing the show attributes, seasons, episodes, banner, etc. `Is working, except for the banner.`
- [x] Episode titles / summaries (episode titles not yet visible, small issue)
- [x] Poster and Banners are downloaded. But for now i'm using the 'original' quality. Meaning in some cases, very large images. Maybe we should decide on a different image size.
- [x] Add metadata creation
- [x] Add banner and poster support if available
- [x] Add actors information (Used for metadata creation)
- [x] ~~Add multilingual support.~~ Currently not supported by indexer.
- [x] Add timezone and network information
- [x] Add indexer mappings. On an indexers level, all external id's need to be mapped to indexer[showid]['externals']. Keys should align with the the indexer_config's 'mapped_to' key.
- [x] Add the indexer icon
- [x] Implement a new method for show updating, as we cannot use the old tvdb xml's anymore. (well really need to test this.
- [ ] Cleanup

### tvmaze Apiv2 Lib related
- [x] Make sure the api returns unicode. Test by searching: 'guide to pretty'. Currently i've adjusted the mapping for the Series modal. Need to also check for others, like episode, actors, etc.
- [x] Pass the indexer requests object. This will improve caching.
- [x] Add searching with apostophes '.

### showUpdates related
More info: https://github.com/pymedusa/Medusa/wiki/Show-indexer-updating
For this the following needs to be implemented:
- [x] Support for force full update
- [x] Update only seasons that reached their next_update.
- [x] Add new shows, should set the next_updates
- [x] Episodes that start a new season, should update the cache indexer_update table with the next_update.

### Externals
Implemented new feature, that maps all external id's for each indexer to the database in indexer_mapping.

The configuration is located in indexer/indexer_config.py.
The methods for loading from indexer and loading from db are located in tv.py.
It can be utilized by calling the externals attribute on a show object. For example. show_object.get('tvdb_id') will give back the tvdb_id if available.
- [x] Added support for searching newznab providers who support tvmaze or tmdb indexer id's.
- [x] Advance support for newznab, using the t=caps: ```<tv-search available="yes" supportedParams="q,vid,tvdbid,traktid,rid,tvmazeid,imdbid,tmdbid,season,ep"/>``` 
Currently we following this proces for using a tvsearch id, like tvdbid or tvmazeid.
* If a newznab indexer does not have supportParams, default to search_string right.
* If a newznab indexer has support for tvdbid (no matter which indexer I use) use tvdbid.
* If a don't have a tvdbid, see if I can use the shows indexers primary id.
* If we don't have a match there, try if we can use an external id.
* Finally also fallback to search_string.

For a list of valid keys, mappings and reverse_mappings can be called from indexer_config.

### Known issues:
- [x] Searching is not possible for indexers that require a tvdbid. Test with: BTN, newznab, hdbits and rarbg.
- [x] Results from a tvmaze search, don't link to the corresponding show on tvmaze.com.
- [x] Searching newznab using daily search results in exception, as there is no specific show to get indexerid from. https://github.com/pymedusa/Medusa/pull/1534/commits/16307e9663c3e888006b06f38cd56d448c7948b6
- [ ] Remove the cause for this warning at startup: `No handlers could be found for logger "knowit.providers.mediainfo"`

### Screenies
image

### CODE REVIEW

File: medusa/indexers/indexer_config.py [hunk 1]
```diff
@@ -41,10 +41,6 @@
 EXTERNAL_IMDB = 10
 EXTERNAL_ANIDB = 11
 
-# mapping = {
```
Comment (p0psicles) [inline medusa/indexers/indexer_config.py]: @OmgImAlexis @ratoaq2 I know I previously provided you with these dicts. But I kinda reused them here for the external id mappings. I've looked where you used it, and implemented alternative solutions. But maybe i've missed something. Better check this.

File: medusa/tv.py [hunk 2]
```diff
@@ -2564,7 +2568,7 @@ def __str__(self):
 
     def to_json(self, detailed=True):
         """Return the json representation."""
-        indexer_name = reverse_mapping[self.indexer]
+        indexer_name = indexerConfig[self.indexer].name
```
Comment (p0psicles) [inline medusa/tv.py]: This is another location, where I assume you want to use the indexers name.

**Comment by p0psicles** [comment]
Grrr.

```
2016-11-20 21:26:27 DEBUG    SEARCHQUEUE-DAILY-SEARCH :: [6box.me] :: [921b9c7] Traceback (most recent call last):
  File "/home/opt/medusa/medusa/search/queue.py", line 248, in run
    found_results = searchForNeededEpisodes()
  File "/home/opt/medusa/medusa/search/core.py", line 446, in searchForNeededEpisodes
    cur_provider.cache.updateCache()
  File "/home/opt/medusa/medusa/tv_cache.py", line 190, in updateCache
    data = self._getRSSData()
  File "/home/opt/medusa/medusa/tv_cache.py", line 176, in _getRSSData
    return {'entries': self.provider.search(self.search_params)} if self.search_params else None
  File "/home/opt/medusa/medusa/providers/nzb/newznab.py", line 95, in search
    match_indexer = self._match_indexer()
  File "/home/opt/medusa/medusa/providers/nzb/newznab.py", line 348, in _match_indexer
    for search_type in self.cap_tv_search.split(','):
AttributeError: 'NoneType' object has no attribute 'split'
```

**Comment by duramato** [comment]
```
2016-11-20 21:07:40 INFO     SEARCHQUEUE-MANUAL-22833 :: [drunkenslug] :: [ae566b4] Performing episode search for Formula 1
2016-11-20 21:07:43 DEBUG    SEARCHQUEUE-MANUAL-22833 :: [drunkenslug] :: [ae566b4] Search mode: Episode
2016-11-20 21:07:43 DEBUG    SEARCHQUEUE-MANUAL-22833 :: [drunkenslug] :: [ae566b4] Search show using search string: indexer_id (tvdb):
2016-11-20 21:07:42 DEBUG    SEARCHQUEUE-MANUAL-22833 :: [drunkenslug] :: [ae566b4] GET URL: https://drunkenslug.com/api?apikey=**********&t=caps [Status: 200]
2016-11-20 21:07:43 DEBUG    SEARCHQUEUE-MANUAL-22833 :: [drunkenslug] :: [ae566b4] GET URL: https://drunkenslug.com/api?apikey=**********&tvdbid=&season=2016&maxage=3000&cat=5000%2C5080%2C5040%2C5050%2C5030%2C5060&limit=100&t=tvsearch&offset=0&tvmazeid=3&ep=11%2F13 [Status: 200]
2016-11-20 21:07:43 INFO     SEARCHQUEUE-MANUAL-22833 :: [drunkenslug] :: [ae566b4] Incorrect parameter (tvdbid must not be empty)
2016-11-20 21:07:43 DEBUG    SEARCHQUEUE-MANUAL-22833 :: [drunkenslug] :: [ae566b4] Fallback season pack search initiate
2016-11-20 21:07:43 INFO     SEARCHQUEUE-MANUAL-22833 :: [drunkenslug] :: [ae566b4] Performing season pack search for Formula 1
```

Searching for a show for example Formula 1 from tvmaze , which doesn't have a tvdb id doesn't try a string search

**Comment by fernandog** [comment]
"Search show using search string" ?

**Comment by duramato** [comment]
`	(invalid Timeformat) Channel 4`
Show's page:
image
Manual search page:
image

**Comment by duramato** [comment]
https://github.com/pymedusa/Medusa/issues/1622

And https://github.com/pymedusa/Medusa/pull/1534#issuecomment-261967077 fix didn't work.

File: medusa/providers/nzb/newznab.py [hunk 3]
```diff
@@ -324,6 +331,42 @@ def image_name(self):
...
+        map_caps = {INDEXER_TMDB: 'tmdbid', INDEXER_TVDBV2: 'tvdbid', INDEXER_TVMAZE: 'tvmazeid'}
+
+        return_mapping = {}
+
+        if self.cap_tv_search == 'True':
+            # We didn't get back a supportedParams, lets return, and continue with doing a search string search.
+            return {}
+
+        for search_type in self.cap_tv_search.split(','):
+            if search_type == 'tvdbid':
```
Comment (duramato) [inline medusa/providers/nzb/newznab.py]: Maybe do `if search_type == 'tvdbid' and self._get_tvdb_id():`
for when there isn't a tvdbid but the indexer supports it

Comment (p0psicles) [inline medusa/providers/nzb/newznab.py]: We already do this, just below this code.

**Comment by p0psicles** [comment]
about #1534, can't reproduce. So you will have to fix that yourself.

**Pull request merged**